lib/service/discovery_test.go

```diff
@@ -70,3 +75,104 @@ func TestTeleportProcessIntegrationsOnly(t *testing.T) {
 		})
 	}
 }
+
+func TestTeleportProcess_initDiscoveryService(t *testing.T) {
+
+	tests := []struct {
+		name      string
+		cfg       servicecfg.AccessGraphConfig
+		rsp       *clusterconfigpb.AccessGraphConfig
+		err       error
+		want      discovery.AccessGraphConfig
+		assertErr require.ErrorAssertionFunc
+	}{
+		{
+			name: "local access graph",
+			cfg: servicecfg.AccessGraphConfig{
+				Enabled:  true,
+				Addr:     "localhost:5000",
+				Insecure: true,
+			},
+			rsp: nil,
+			err: nil,
+			want: discovery.AccessGraphConfig{
+				Enabled:  true,
+				Addr:     "localhost:5000",
+				Insecure: true,
+			},
+			assertErr: require.NoError,
+		},
+		{
+			name: "access graph disabled locally but enabled in auth",
+			cfg: servicecfg.AccessGraphConfig{
+				Enabled: false,
+			},
+			rsp: &clusterconfigpb.AccessGraphConfig{
+				Enabled:  true,
+				Address:  "localhost:5000",
+				Insecure: true,
+			},
+			err: nil,
+			want: discovery.AccessGraphConfig{
+				Enabled:  true,
+				Addr:     "localhost:5000",
+				Insecure: true,
+			},
+			assertErr: require.NoError,
+		},
+		{
+			name: "access graph disabled locally and auth doesn't implement GetClusterAccessGraphConfig",
+			cfg: servicecfg.AccessGraphConfig{
+				Enabled: false,
+			},
+			rsp: nil,
+			err: trace.NotImplemented("err"),
+			want: discovery.AccessGraphConfig{
+				Enabled: false,
+			},
+			assertErr: require.NoError,
+		},
+		{
+			name: "access graph disabled locally and auth call fails",
+			cfg: servicecfg.AccessGraphConfig{
+				Enabled: false,
+			},
+			rsp: nil,
+			err: trace.BadParameter("err"),
+			want: discovery.AccessGraphConfig{
+				Enabled: false,
+			},
+			assertErr: require.Error,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			accessGraphCfg, err := buildAccessGraphFromTAGOrFallbackToAuth(
+				context.Background(),
+				&servicecfg.Config{
+					AccessGraph: tt.cfg,
+				},
+				&fakeClient{
+					rsp: tt.rsp,
+					err: tt.err,
+				},
+				slog.Default(),
+			)
+			tt.assertErr(t, err)
+			require.Equal(t, tt.want, accessGraphCfg)
```

Comment:
When an error is returned, the returned value should not matter.

Reply:
it's comparing the error with nil so it's just a way of checking if err!=nil{return}